Fix bulk remove in Local Collection to remove all items (#13959)

DipakHalkude opened the pull request:
Context: https://github.com/meteor/meteor/issues/13959

DipakHalkude committed:
Fix bulk remove in Local Collection to remove all items (#13959)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/meteor/meteor?pullRequest=13965) <br/>All committers have signed the CLA.

nachocodoner commented:
Ideally, we should have a test in Minimongo that demonstrates the regression you reported. It should show how the issue occurs when iterating and removing only one item, and confirm the test fails when expecting all items to be removed. Your fix here will show how that test solves to green.

Could you create a simple test for that?

https://github.com/meteor/meteor/blob/devel/DEVELOPMENT.md#tests

italojs commented:
duplicated https://github.com/meteor/meteor/pull/13965

DipakHalkude commented:
> Ideally, we should have a test in Minimongo that demonstrates the regression you reported. It should show how the issue occurs when iterating and removing only one item, and confirm the test fails when expecting all items to be removed. Your fix here will show how that test solves to green.
> 
> Could you create a simple test for that?
> 
> https://github.com/meteor/meteor/blob/devel/DEVELOPMENT.md#tests

I am on it :) Please wait I am adding the test cases :)

DipakHalkude committed:
Fix bulk remove in LocalCollection to remove all matching documents
- Change condition from !fn(doc, id) to fn(doc, id) === false
- This prevents early loop termination when callback returns undefined
- Add test case to verify bulk remove with  operator works correctly
- Fixes #13959 where only one item was removed in bulk operations

DipakHalkude committed:
Update local_collection.js

DipakHalkude committed:
Delete packages/minimongo/minimongo_tests_client.js

DipakHalkude committed:
Fix bulk remove in LocalCollection to remove all matching documents

- Change condition from !fn(doc, id) to fn(doc, id) === false
- This prevents early loop termination when callback returns undefined
- Add test case to verify bulk remove with  operator works correctly
- Fixes #13959 where only one item was removed in bulk operations

DipakHalkude committed:
Fix bulk remove in LocalCollection to remove all matching documents

- Change condition from !fn(doc, id) to fn(doc, id) === false
- This prevents early loop termination when callback returns undefined
- Add test case to verify bulk remove with  operator works correctly
- Fixes #13959 where only one item was removed in bulk operations

DipakHalkude commented:
Hi @nachocodoner I’ve added a simple test in Minimongo that reproduces the regression related to iterating and removing a single item. The test initially fails when expecting all items to be removed, and with the fix applied, it now passes successfully — confirming the issue is resolved.

All existing tests are also passing. Please have a look and let me know if any additional adjustments are needed.
Thank you :)

nachocodoner commented on packages/minimongo/minimongo_tests.js:
```diff
@@ -59,6 +59,36 @@ Tinytest.add('minimongo - wrapTransform', test => {
+  
+  // Verify the correct documents were removed
+  test.isNull(coll.findOne('id1'));
+  test.isNull(coll.findOne('id2'));
```
I think here it will be isUndefined instead of null, to pass properly.

DipakHalkude replied:
Ok I will just change it to undefined and commit the changes :) Is it fine then ?

nachocodoner replied:
Yes, locally, with isNull, tests were failing for me when doing `./meteor test-packages minimongo`. Did it for you?

DipakHalkude replied:
No for me that working fine. I guess there is other problem but changing from ```isNull``` to ```isUndefined``` will not cause any error. It is fine :)

DipakHalkude replied:
Can I just update that and commit it?

nachocodoner replied:
Yes, please, lets see if CI will also go green later. 😄

DipakHalkude replied:
Ok :)

DipakHalkude replied:
Just have a clarification : 

     test.isUndefined(coll.findOne('id1'));
     test.isUndefined(coll.findOne('id2'));
     test.isNotNull(coll.findOne('id3'));
     test.isNotNull(coll.findOne('id4'));

OR

     test.isUndefined(coll.findOne('id1'));
     test.isUndefined(coll.findOne('id2'));
     test.isNotUndefined(coll.findOne('id3'));
     test.isNotUndefined(coll.findOne('id4'));

nachocodoner commented:
Confirmed the fix solves the issue:

<img width="720" height="467" alt="image" src="https://github.com/user-attachments/assets/b892db5e-4510-4cfd-8f9b-9a239fb6503f" />


The only issue I got with the tests is that I had to compare using isUndefined instead of isNull.

nachocodoner commented:
@DipakHalkude: What about to `return true` in the place that the boolean is missing, https://github.com/meteor/meteor/pull/13964/files#diff-de17c33eebf4b8da6bff1f392c3f4b9e54da4265dd262361bafebcfbe49da10eR268? as @logiclrd tried on his PR

This way it may result in a better fix as we keep boolean expectation when use `_eachPossiblyMatchingDocSync`.

DipakHalkude commented:
@nachocodoner  I believe my current fix is more robust because it addresses the root issue in the ```_eachPossiblyMatchingDocSync``` method itself. Changing the condition to only break when explicitly returning ```false```  ensures consistent behavior across all uses of this method, not just the remove operation. The alternative approach would require modifying every callback that uses this method to return ```true``` , which could be error-prone.

nachocodoner reviewed (dismissed):
Ok, good, yes, I am going to get your fix as the final one.

[Lodash's forEach](https://lodash.com/docs/4.17.15#forEach) behaves similarly, so it will only skip the loop if you explicitely return `false` as your fix intended. Being less error-prone when using the iterator.

Thank you for your contribution 🚀, lets wait the tests signals to go green.

logiclrd commented:
> The alternative approach would require modifying every callback that uses this method to return `true` , which could be error-prone.

Devil's advocate: There is only one other use in the entire codebase, and it already explicitly returns `true`/`false` to control enumeration. But I see your argument that this is more robust. :-)

logiclrd commented:
By the way @DipakHalkude, you created this PR without checking if there was _already_ a PR on this exact topic. It would have been appropriate to join the discussion on the existing PR rather than simply code up an independent solution and make a competing PR. If granted permission to my fork of the repo, you could have made commits there directly, too, resulting in a co-authored contribution.

DipakHalkude committed:
Fix test assertions: use isUndefined/isNotUndefined instead of isNull/isNotNull

DipakHalkude commented:
Thanks for pointing that out, @logiclrd — I totally understand and appreciate the reminder.
I admit I jumped straight into fixing it after analyzing the issue and didn’t notice your existing PR at first.
I’ll make sure to check for open discussions or PRs next time before creating a new one.

Your earlier work and context were really helpful in understanding the regression, and I’d be glad to collaborate directly in the future if there’s a similar case. Thanks again for the feedback and for reviewing the fix! 🙏

DipakHalkude commented:
Thank you so much for your guidance and support throughout this fix, @nachocodoner 🙏
I’ve made the requested updates — everything’s done from my side and all tests are passing successfully ✅
Whenever you get a moment, please review and approve the PR. Really appreciate your help! 😊

DipakHalkude commented:
Hi @nachocodoner, all checks have passed now, including CI ✅
Everything is done from my side. Thanks so much for your guidance and support throughout! 🙏

nachocodoner commented:
Cool. I’ve marked this PR to be included in the next Meteor 3.4 contributions beta, which will lead to the official release afterward.

Thanks for your work on this and other issues. It’s great to see how precise your fix was, and I encourage you to keep contributing.

Please make sure others aren’t already working on the same issue next time, so we can give everyone a chance to contribute. I’ll also keep this in mind going forward.

DipakHalkude commented:
Thank you so much, @nachocodoner! I really appreciate your kind words and guidance throughout this process. I’m glad the fix worked well and will be part of the upcoming Meteor 3.4 beta. I’ll make sure to check for existing PRs and discussions next time to coordinate better with others. Grateful for the opportunity to contribute and looking forward to more collaborations ahead! 🚀
